### PR TITLE
make: add an end-to-end test

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Must be doing something wrong, blocked on https://github.com/kcp-dev/kcp/issues/1742

File: Makefile [hunk 1]
```diff
@@ -43,15 +43,15 @@ help: ## Display this help.
 ##@ Development
 
 .PHONY: manifests
-manifests: controller-gen ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
+manifests: $(CONTROLLER_GEN) ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
```
Comment (stevekuznetsov) [inline Makefile:46]: nit: with these as phony, they always re-ran for me ...

File: Makefile [hunk 2]
```diff
@@ -63,9 +63,65 @@ vet: ## Run go vet against code.
...
+.PHONY: ready-deployment
+ready-deployment: KUBECONFIG = $(ARTIFACT_DIR)/kcp.kubeconfig
+ready-deployment: kind-image install deploy
+	$(KCP_KUBECTL) --namespace "controller-runtime-example-system" rollout status deployment/controller-runtime-example-controller-manager
+
+.PHONY: kind-image
+kind-image: docker-build
+	kind load docker-image $(IMG) --name controller-runtime-example
+
+$(ARTIFACT_DIR)/kind.kubeconfig: $(ARTIFACT_DIR) require-kind
```
Comment (stevekuznetsov) [inline Makefile:88]: @sttts how to debug why this runs always even when the output file exists would be sweet

File: Makefile [hunk 3]
```diff
@@ -63,9 +63,65 @@ vet: ## Run go vet against code.
...
+KCP_KUBECTL ?= PATH=$(LOCALBIN):$(PATH) KUBECONFIG=$(ARTIFACT_DIR)/kcp.kubeconfig kubectl
+KIND_KUBECTL ?= kubectl --kubeconfig $(ARTIFACT_DIR)/kind.kubeconfig
+
+.PHONY: kcp-synctarget
+kcp-synctarget: require-yq kcp-workspace $(ARTIFACT_DIR)/syncer.yaml
+	$(KIND_KUBECTL) apply -f $(ARTIFACT_DIR)/syncer.yaml
+	$(eval DEPLOYMENT_NAME = $(shell yq 'select(.kind=="Deployment") | .metadata.name' --raw-output < $(ARTIFACT_DIR)/syncer.yaml ))
+	$(eval DEPLOYMENT_NAMESPACE = $(shell yq 'select(.kind=="Deployment") | .metadata.namespace' --raw-output < $(ARTIFACT_DIR)/syncer.yaml ))
+	$(KIND_KUBECTL) --namespace $(DEPLOYMENT_NAMESPACE) rollout status deployment/$(DEPLOYMENT_NAME)
+	@if [[ ! -s $(ARTIFACT_DIR)/syncer.log ]]; then ( $(KIND_KUBECTL) --namespace $(DEPLOYMENT_NAMESPACE) logs deployment/$(DEPLOYMENT_NAME) -f >$(ARTIFACT_DIR)/syncer.log 2>&1 & ); fi
```
Comment (stevekuznetsov) [inline Makefile:104]: We need a `kubectl kcp workload sync logs` or something

File: Makefile [hunk 4]
```diff
@@ -1,3 +1,41 @@
...
+	curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN)
+
+$(CONTROLLER_GEN): $(LOCALBIN)
+	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
+
+OS ?= $(shell go env GOOS )
+ARCH ?= $(shell go env GOARCH )
+
+$(KCP): $(LOCALBIN) ## Download kcp locally if necessary.
+	wget -qO- https://github.com/kcp-dev/kcp/releases/download/v$(KCP_VERSION)/kcp_$(KCP_VERSION)_$(OS)_$(ARCH).tar.gz | tar -xvzf - bin/kcp --directory $(LOCALBIN)/../
```
Comment (ncdc) [inline Makefile:31]: Might want to consider `curl` instead - Macs have that but not wget by default

**Comment by stevekuznetsov** [comment]
Depends on https://github.com/openshift/release/pull/31458#event-7214291845

File: Makefile [hunk 5]
```diff
@@ -1,6 +1,73 @@
...
+$(YQ): ## Download yq locally if necessary.
+	mkdir -p $(LOCALBIN)
+	GOBIN=$(LOCALBIN) go install github.com/mikefarah/yq/v4@$(YQ_VERSION)
+
+OS ?= $(shell go env GOOS )
+ARCH ?= $(shell go env GOARCH )
+
+$(KCP): ## Download kcp locally if necessary.
+	mkdir -p $(LOCALBIN)
+	curl -L -s -o - https://github.com/kcp-dev/kcp/releases/download/v$(KCP_VERSION)/kcp_$(KCP_VERSION)_$(OS)_$(ARCH).tar.gz | tar --directory $(LOCALBIN)/../ -xvzf - bin/kcp
```
Comment (stevekuznetsov) [inline Makefile:56]: @ncdc could you please verify this on macos?

```
rm bin/kcp && make $( realpath bin/kcp )
```

**Comment by stevekuznetsov** [comment]
@ncdc this is ready for review

File: Makefile (hunk 5 above)
Comment (ncdc) [reply Makefile:56]: Mac doesn't have `realpath` OOTB

Comment (stevekuznetsov) [reply Makefile:56]: Er - the target does not use `realpath`, that was just my shell command. Use whatever you want to to get the absolute path to the binary and run that through `make`

Comment (ncdc) [reply Makefile:56]:
```
$ make `pwd`/bin/kcp
mkdir -p /Users/angoldst/code/kcp/controller-runtime-example/bin
curl -L -s -o - https://github.com/kcp-dev/kcp/releases/download/v0.7.5/kcp_0.7.5_darwin_amd64.tar.gz | tar --directory /Users/angoldst/code/kcp/controller-runtime-example/bin/../ -xvzf - bin/kcp
x bin/kcp
touch /Users/angoldst/code/kcp/controller-runtime-example/bin/kcp # we download an "old" file, so make will re-download to refresh it unless we make it newer than the owning dir

$ make `pwd`/bin/kcp
make: `/Users/angoldst/code/kcp/controller-runtime-example/bin/kcp' is up to date.
```

Comment (stevekuznetsov) [reply Makefile:56]: :tada:

**Review by ncdc — APPROVED** [review]

**Pull request merged**